File: laya_dino/game.py
"""Faithful Chrome-dino simulation.

Ported directly from the Chromium t-rex runner (`third_party/t-rex-runner/
index.js`) so that a policy trained here transfers to the real game in a
browser. The first version of this file approximated the game; that was fine
for measuring the decision window but would have mistrained every threshold.
The differences that mattered:

* **Dino x position** is 50, not 25. A 25px shift is 2-4 frames of decision
  time at speed — larger than the margin we're optimising.
* **Jump velocity depends on speed**: `-10 - speed/10`, so a jump at max speed
  is meaningfully higher and longer than one at the start.
* **Per-frame rounding.** The real game does `yPos += Math.round(velocity)`,
  which is not the same trajectory as unrounded float accumulation.
* **Cacti spawn in clusters** of up to 3 (`MAX_OBSTACLE_LENGTH`), gated on
  speed. A 75px-wide obstacle needs an earlier jump than a 25px one, and a
  model that has never seen one will misjudge it.
* **Real collision boxes.** Chromium uses 6 boxes for the dino and 3 per
  cactus rather than one rectangle, which moves the last-safe-jump frame.
* **Gap formula** is `round(width*speed + minGap*0.6)` to 1.5x that, per type.

Two properties are load-bearing and preserved:

* **Determinism.** Same seed + same actions produce an identical replay, which
  is what makes the oracle's labels exact and evaluation reproducible.
* **No wall-clock coupling.** The game advances only when `tick()` is called,
  so a 22ms model decision cannot corrupt the physics.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

# --- Runner.config ---------------------------------------------------------
CANVAS_W = 600.0
CANVAS_H = 150.0
ACCELERATION = 0.001
GAP_COEFFICIENT = 0.6
GRAVITY = 0.6
START_SPEED = 6.0
MAX_SPEED = 13.0
MAX_OBSTACLE_LENGTH = 3
MAX_OBSTACLE_DUPLICATION = 2
BOTTOM_PAD = 10.0
MAX_GAP_COEFFICIENT = 1.5

# --- Trex.config -----------------------------------------------------------
DINO_X = 50.0          # START_X_POS
DINO_W = 44.0
DINO_H = 47.0
INITIAL_JUMP_VELOCITY = -10.0
GROUND_Y = CANVAS_H - DINO_H - BOTTOM_PAD   # 93.0

# The jump is *capped*, which matters more than anything else in this file.
# Chromium's `updateJump` calls `endJump()` once `yPos < MAX_JUMP_HEIGHT`, and
# `endJump` clamps the velocity to DROP_VELOCITY, cutting the ascent short.
# Without this the dino appears to follow a full parabola and reach ~94px,
# when the real game tops out far lower. Modelling the uncapped arc made the
# sim believe a 51px cactus cluster was clearable from 201px away; in the real
# browser that same jump hits the cactus. Every trained threshold inherits
# this, so it has to be exact.
MAX_JUMP_HEIGHT = 30.0      # absolute yPos, not a height above ground
MIN_JUMP_HEIGHT = 30.0      # offset below groundY at which the jump may be cut
DROP_VELOCITY = -5.0
MIN_JUMP_Y = GROUND_Y - MIN_JUMP_HEIGHT     # 63.0

SCORE_COEFFICIENT = 0.025

# Trex.collisionBoxes.RUNNING — (x, y, w, h) relative to the dino's top-left.
TREX_BOXES = (
    (22.0, 0.0, 17.0, 16.0),
    (1.0, 18.0, 30.0, 9.0),
    (10.0, 35.0, 14.0, 8.0),
    (1.0, 24.0, 29.0, 5.0),
    (5.0, 30.0, 21.0, 4.0),
    (9.0, 34.0, 15.0, 4.0),
)

CACTUS_SMALL = "cactus_small"
CACTUS_LARGE = "cactus_large"
PTERODACTYL = "pterodactyl"

# Obstacle.types, keyed by our names.
TYPES = {
    CACTUS_SMALL: {
        "width": 17.0, "height": 35.0, "y_pos": (105.0,),
        "multiple_speed": 4.0, "min_gap": 120.0, "min_speed": 0.0,
        "boxes": ((0.0, 7.0, 5.0, 27.0), (4.0, 0.0, 6.0, 34.0), (10.0, 4.0, 7.0, 14.0)),
    },
    CACTUS_LARGE: {
        "width": 25.0, "height": 50.0, "y_pos": (90.0,),
        "multiple_speed": 7.0, "min_gap": 120.0, "min_speed": 0.0,
        "boxes": ((0.0, 12.0, 7.0, 38.0), (8.0, 0.0, 7.0, 49.0), (13.0, 10.0, 10.0, 38.0)),
    },
    PTERODACTYL: {
        "width": 46.0, "height": 40.0, "y_pos": (100.0, 75.0, 50.0),
        "multiple_speed": 999.0, "min_gap": 150.0, "min_speed": 8.5,
        "boxes": ((15.0, 15.0, 16.0, 5.0), (18.0, 21.0, 24.0, 6.0),
                  (2.0, 14.0, 4.0, 3.0), (6.0, 10.0, 4.0, 7.0), (10.0, 8.0, 6.0, 9.0)),
    },
}
TYPE_ORDER = (CACTUS_SMALL, CACTUS_LARGE, PTERODACTYL)

RUN = 0
JUMP = 1

_MASK64 = (1 << 64) - 1


class Rng:
    """PCG-XSH-RR 32. Deterministic and dependency-free.

    The real game uses `Math.random()`, which we cannot reproduce — so levels
    here are *statistically* like Chrome's, not identical to any particular
    Chrome session. That's fine: the policy must generalise over levels
    anyway, and determinism is what we need for labelling.
    """

    __slots__ = ("state",)

    def __init__(self, seed: int):
        self.state = (seed * 6364136223846793005 + 1) & _MASK64
        self.next_u32()

    def clone(self) -> "Rng":
        r = Rng.__new__(Rng)
        r.state = self.state
        return r

    def next_u32(self) -> int:
        old = self.state
        self.state = (old * 6364136223846793005 + 1442695040888963407) & _MASK64
        xorshifted = (((old >> 18) ^ old) >> 27) & 0xFFFFFFFF
        rot = (old >> 59) & 31
        return ((xorshifted >> rot) | (xorshifted << (32 - rot))) & 0xFFFFFFFF

    def next_f32(self) -> float:
        return (self.next_u32() >> 8) / float(1 << 24)

    def randint(self, lo: int, hi: int) -> int:
        """Inclusive, matching the game's `getRandomNum(min, max)`."""
        return lo + self.next_u32() % (hi - lo + 1)


@dataclass(slots=True)
class Obstacle:
    kind: str
    size: int
    x: float
    y: float
    w: float
    h: float
    gap: float
    boxes: tuple
    following_created: bool = False

    def distance_from_dino(self) -> float:
        return self.x - (DINO_X + DINO_W)


@dataclass(slots=True)
class Observation:
    """Everything a controller may see, and exactly what gets serialized for
    the model. Keep it small: token count dominates inference latency."""

    distance: float
    speed: float
    obstacle: Optional[str]
    obstacle_w: float
    obstacle_h: float
    obstacle_ground_offset: float
    size: int
    airborne: bool
    dino_y_velocity: float
    frames_to_impact: float


def _boxes_for(kind: str, size: int, width: float) -> tuple:
    """Collision boxes, widened for clustered cacti exactly as the game does:
    the middle box stretches and the right box slides to the new edge."""
    b = [list(x) for x in TYPES[kind]["boxes"]]
    if size > 1:
        b[1][2] = width - b[0][2] - b[2][2]
        b[2][0] = width - b[2][2]
    return tuple(tuple(x) for x in b)


class Game:
    __slots__ = (
        "rng", "speed", "distance_ran", "frame", "dino_y", "dino_vy",
        "jumping", "obstacles", "crashed", "history", "reached_min_height",
    )

    def __init__(self, seed: int):
        self.rng = Rng(seed)
        self.speed = START_SPEED
        self.distance_ran = 0.0
        self.frame = 0
        self.dino_y = GROUND_Y
        self.dino_vy = 0.0
        self.jumping = False
        self.reached_min_height = False
        self.obstacles: list[Obstacle] = []
        self.crashed = False
        self.history: list[str] = []
        self._add_obstacle()

    def clone(self) -> "Game":
        """Fast fork for oracle lookahead; `copy.deepcopy` is ~20x slower and
        this runs tens of millions of times during dataset generation."""
        g = Game.__new__(Game)
        g.rng = self.rng.clone()
        g.speed = self.speed
        g.distance_ran = self.distance_ran
        g.frame = self.frame
        g.dino_y = self.dino_y
        g.dino_vy = self.dino_vy
        g.jumping = self.jumping
        g.reached_min_height = self.reached_min_height
        g.obstacles = [
            Obstacle(o.kind, o.size, o.x, o.y, o.w, o.h, o.gap, o.boxes,
                     o.following_created)
            for o in self.obstacles
        ]
        g.crashed = self.crashed
        g.history = list(self.history)
        return g

    def score(self) -> int:
        return int(self.distance_ran * SCORE_COEFFICIENT)

    def at_max_speed(self) -> bool:
        return self.speed >= MAX_SPEED

    def _next_obstacle(self) -> Optional[Obstacle]:
        pending = [o for o in self.obstacles if o.x + o.w > DINO_X]
        return min(pending, key=lambda o: o.x) if pending else None

    def observe(self) -> Observation:
        o = self._next_obstacle()
        if o is None:
            return Observation(math.inf, self.speed, None, 0.0, 0.0, 0.0, 0,
                               self.jumping, self.dino_vy, math.inf)
        d = o.distance_from_dino()
        return Observation(
            distance=d,
            speed=self.speed,
            obstacle=o.kind,
            obstacle_w=o.w,
            obstacle_h=o.h,
            # Height of the obstacle's bottom edge above the ground line.
            obstacle_ground_offset=(GROUND_Y + DINO_H) - (o.y + o.h),
            size=o.size,
            airborne=self.jumping,
            dino_y_velocity=self.dino_vy,
            frames_to_impact=d / self.speed,
        )

    def _duplicate_check(self, kind: str) -> bool:
        return (len(self.history) >= MAX_OBSTACLE_DUPLICATION
                and all(h == kind for h in self.history[:MAX_OBSTACLE_DUPLICATION]))

    def _add_obstacle(self) -> None:
        # The game retries type selection until one is allowed; bound the loop
        # so a pathological state can't hang the simulator.
        for _ in range(32):
            kind = TYPE_ORDER[self.rng.randint(0, len(TYPE_ORDER) - 1)]
            cfg = TYPES[kind]
            if self._duplicate_check(kind) or self.speed < cfg["min_speed"]:
                continue
            break
        else:
            kind, cfg = CACTUS_SMALL, TYPES[CACTUS_SMALL]

        size = self.rng.randint(1, MAX_OBSTACLE_LENGTH)
        if size > 1 and cfg["multiple_speed"] > self.speed:
            size = 1
        width = cfg["width"] * size

        y_opts = cfg["y_pos"]
        y = y_opts[self.rng.randint(0, len(y_opts) - 1)] if len(y_opts) > 1 else y_opts[0]

        min_gap = round(width * self.speed + cfg["min_gap"] * GAP_COEFFICIENT)
        max_gap = round(min_gap * MAX_GAP_COEFFICIENT)
        gap = float(self.rng.randint(int(min_gap), int(max_gap)))

        self.obstacles.append(Obstacle(
            kind=kind, size=size, x=CANVAS_W, y=y, w=width, h=cfg["height"],
            gap=gap, boxes=_boxes_for(kind, size, width),
        ))
        self.history.insert(0, kind)
        del self.history[MAX_OBSTACLE_DUPLICATION:]

    def _collides(self) -> bool:
        """Chromium's two-stage check: cheap outer-box test, then the detailed
        per-sprite boxes. Using one rectangle instead shifts the last-safe-jump
        frame, which is exactly the boundary we train on."""
        dx, dy = DINO_X, self.dino_y
        for o in self.obstacles:
            # Outer box first.
            if not (dx < o.x + o.w and dx + DINO_W > o.x
                    and dy < o.y + o.h and dy + DINO_H > o.y):
                continue
            for tb in TREX_BOXES:
                tx0, ty0 = dx + tb[0], dy + tb[1]
                tx1, ty1 = tx0 + tb[2], ty0 + tb[3]
                for ob in o.boxes:
                    ox0, oy0 = o.x + ob[0], o.y + ob[1]
                    ox1, oy1 = ox0 + ob[2], oy0 + ob[3]
                    if tx0 < ox1 and tx1 > ox0 and ty0 < oy1 and ty1 > oy0:
                        return True
        return False

    def tick(self, action: int) -> bool:
        """Advance exactly one frame. Returns False once the run has ended."""
        if self.crashed:
            return False

        if action == JUMP and not self.jumping:
            self.jumping = True
            self.reached_min_height = False
            # Chromium: jumpVelocity = INITIAL_JUMP_VELOCITY - (speed / 10)
            self.dino_vy = INITIAL_JUMP_VELOCITY - (self.speed / 10.0)

        if self.jumping:
            # Chromium rounds the per-frame displacement.
            self.dino_y += round(self.dino_vy)
            self.dino_vy += GRAVITY

            if self.dino_y < MIN_JUMP_Y:
                self.reached_min_height = True
            # Reached max height -> endJump(): clamp the ascent. This is what
            # makes the real jump much shorter than a free parabola.
            if self.dino_y < MAX_JUMP_HEIGHT:
                if self.reached_min_height and self.dino_vy < DROP_VELOCITY:
                    self.dino_vy = DROP_VELOCITY

            if self.dino_y > GROUND_Y:
                self.dino_y = GROUND_Y
                self.dino_vy = 0.0
                self.jumping = False
                self.reached_min_height = False

        for o in self.obstacles:
            o.x -= self.speed
        self.obstacles = [o for o in self.obstacles if o.x + o.w > 0]

        if self.obstacles:
            last = self.obstacles[-1]
            if (not last.following_created
                    and last.x + last.w + last.gap < CANVAS_W):
                self._add_obstacle()
                last.following_created = True
        else:
            self._add_obstacle()

        if self._collides():
            self.crashed = True
            return False

        self.distance_ran += self.speed
        if self.speed < MAX_SPEED:
            self.speed += ACCELERATION
        self.frame += 1
        return True

File: spike/rust-sim/src/game.rs
//! Deterministic Chrome-dino simulation.
//!
//! Constants follow the Chromium t-rex runner (`offline/`), so behaviour and
//! difficulty ramp match the real game. The sim is a pure fixed-timestep state
//! machine: no clock, no rendering, no I/O. One `tick` == one frame at 60fps.
//! Given the same seed and the same action sequence it always replays
//! identically, which is what makes it usable for dataset generation and for
//! decoupling the game clock from inference latency.

pub const CANVAS_W: f32 = 600.0;
pub const CANVAS_H: f32 = 150.0;

pub const GRAVITY: f32 = 0.6;
pub const INITIAL_JUMP_VELOCITY: f32 = -10.0;
pub const START_SPEED: f32 = 6.0;
pub const MAX_SPEED: f32 = 13.0;
pub const ACCELERATION: f32 = 0.001;

pub const DINO_X: f32 = 25.0;
pub const DINO_W: f32 = 44.0;
pub const DINO_H: f32 = 47.0;
pub const GROUND_Y: f32 = CANVAS_H - DINO_H - 10.0;

/// Chromium multiplies distance by this to get the displayed score.
pub const SCORE_COEFFICIENT: f32 = 0.025;

/// Collision boxes are inset slightly, mirroring the real game's forgiving
/// per-sprite hitboxes without modelling each one.
const HITBOX_INSET: f32 = 4.0;

const MIN_GAP_COEFFICIENT: f32 = 0.6;
const MAX_GAP_COEFFICIENT: f32 = 1.5;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ObstacleKind {
    CactusSmall,
    CactusLarge,
    Pterodactyl,
}

impl ObstacleKind {
    pub fn size(self) -> (f32, f32) {
        match self {
            ObstacleKind::CactusSmall => (17.0, 35.0),
            ObstacleKind::CactusLarge => (25.0, 50.0),
            ObstacleKind::Pterodactyl => (46.0, 40.0),
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            ObstacleKind::CactusSmall => "cactus_small",
            ObstacleKind::CactusLarge => "cactus_large",
            ObstacleKind::Pterodactyl => "pterodactyl",
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Obstacle {
    pub kind: ObstacleKind,
    pub x: f32,
    pub y: f32,
    pub w: f32,
    pub h: f32,
}

impl Obstacle {
    /// Horizontal gap between the dino's leading edge and this obstacle.
    /// Negative once the dino has passed it.
    pub fn distance_from_dino(&self) -> f32 {
        self.x - (DINO_X + DINO_W)
    }
}

/// Small deterministic PRNG (PCG-XSH-RR 32). Avoids a dependency and
/// guarantees byte-identical replays across platforms and Rust versions.
#[derive(Clone, Copy, Debug)]
pub struct Rng {
    state: u64,
}

impl Rng {
    pub fn new(seed: u64) -> Self {
        let mut r = Rng { state: seed.wrapping_mul(6364136223846793005).wrapping_add(1) };
        r.next_u32();
        r
    }

    pub fn next_u32(&mut self) -> u32 {
        let old = self.state;
        self.state = old
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        let xorshifted = (((old >> 18) ^ old) >> 27) as u32;
        let rot = (old >> 59) as u32;
        xorshifted.rotate_right(rot)
    }

    /// Uniform in [0, 1).
    pub fn next_f32(&mut self) -> f32 {
        (self.next_u32() >> 8) as f32 / (1u32 << 24) as f32
    }

    pub fn range(&mut self, lo: f32, hi: f32) -> f32 {
        lo + (hi - lo) * self.next_f32()
    }
}

/// The action the controller may take on a given frame.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Action {
    Run,
    Jump,
}

/// Everything a controller is allowed to see. This is also exactly what gets
/// serialized and handed to the model, so it must stay small — token count is
/// the dominant term in inference latency.
#[derive(Clone, Copy, Debug)]
pub struct Observation {
    /// Pixels between the dino and the next obstacle. `f32::INFINITY` if none.
    pub distance: f32,
    pub speed: f32,
    pub obstacle: Option<ObstacleKind>,
    pub obstacle_w: f32,
    pub obstacle_h: f32,
    /// Height of the obstacle's bottom edge above ground (pterodactyls fly).
    pub obstacle_ground_offset: f32,
    pub airborne: bool,
    pub dino_y_velocity: f32,
    /// Frames until the obstacle reaches the dino at the current speed.
    pub frames_to_impact: f32,
}

#[derive(Clone, Debug)]
pub struct Game {
    pub rng: Rng,
    pub speed: f32,
    pub distance_ran: f32,
    pub frame: u64,
    pub dino_y: f32,
    pub dino_vy: f32,
    pub jumping: bool,
    pub obstacles: Vec<Obstacle>,
    pub crashed: bool,
    next_spawn_gap: f32,
}

impl Game {
    pub fn new(seed: u64) -> Self {
        let mut g = Game {
            rng: Rng::new(seed),
            speed: START_SPEED,
            distance_ran: 0.0,
            frame: 0,
            dino_y: GROUND_Y,
            dino_vy: 0.0,
            jumping: false,
            obstacles: Vec::with_capacity(8),
            crashed: false,
            next_spawn_gap: 0.0,
        };
        g.spawn_obstacle();
        g
    }

    pub fn score(&self) -> u32 {
        (self.distance_ran * SCORE_COEFFICIENT) as u32
    }

    pub fn at_max_speed(&self) -> bool {
        self.speed >= MAX_SPEED
    }

    /// Obstacles the dino has not yet cleared, nearest first.
    fn next_obstacle(&self) -> Option<&Obstacle> {
        self.obstacles
            .iter()
            .filter(|o| o.x + o.w > DINO_X)
            .min_by(|a, b| a.x.partial_cmp(&b.x).unwrap())
    }

    pub fn observe(&self) -> Observation {
        match self.next_obstacle() {
            Some(o) => Observation {
                distance: o.distance_from_dino(),
                speed: self.speed,
                obstacle: Some(o.kind),
                obstacle_w: o.w,
                obstacle_h: o.h,
                obstacle_ground_offset: (GROUND_Y + DINO_H) - (o.y + o.h),
                airborne: self.jumping,
                dino_y_velocity: self.dino_vy,
                frames_to_impact: o.distance_from_dino() / self.speed,
            },
            None => Observation {
                distance: f32::INFINITY,
                speed: self.speed,
                obstacle: None,
                obstacle_w: 0.0,
                obstacle_h: 0.0,
                obstacle_ground_offset: 0.0,
                airborne: self.jumping,
                dino_y_velocity: self.dino_vy,
                frames_to_impact: f32::INFINITY,
            },
        }
    }

    fn spawn_obstacle(&mut self) {
        // Pterodactyls only appear later, as in the real game.
        let kind = if self.score() > 450 && self.rng.next_f32() < 0.2 {
            ObstacleKind::Pterodactyl
        } else if self.rng.next_f32() < 0.5 {
            ObstacleKind::CactusLarge
        } else {
            ObstacleKind::CactusSmall
        };
        let (w, h) = kind.size();

        // Pterodactyls spawn at one of three altitudes; the lowest is jumpable,
        // the highest can be run under.
        let ground_offset = if kind == ObstacleKind::Pterodactyl {
            [0.0, 25.0, 50.0][(self.rng.next_u32() % 3) as usize]
        } else {
            0.0
        };
        let y = (GROUND_Y + DINO_H) - h - ground_offset;

        self.obstacles.push(Obstacle { kind, x: CANVAS_W, y, w, h });

        let gap_base = w * self.speed + 40.0 * self.speed;
        self.next_spawn_gap = gap_base * self.rng.range(MIN_GAP_COEFFICIENT, MAX_GAP_COEFFICIENT);
    }

    fn collides(&self) -> bool {
        let dx0 = DINO_X + HITBOX_INSET;
        let dx1 = DINO_X + DINO_W - HITBOX_INSET;
        let dy0 = self.dino_y + HITBOX_INSET;
        let dy1 = self.dino_y + DINO_H - HITBOX_INSET;
        self.obstacles.iter().any(|o| {
            let ox0 = o.x + HITBOX_INSET;
            let ox1 = o.x + o.w - HITBOX_INSET;
            let oy0 = o.y + HITBOX_INSET;
            let oy1 = o.y + o.h - HITBOX_INSET;
            dx0 < ox1 && dx1 > ox0 && dy0 < oy1 && dy1 > oy0
        })
    }

    /// Advance exactly one frame. Returns `false` once the run has ended.
    pub fn tick(&mut self, action: Action) -> bool {
        if self.crashed {
            return false;
        }

        if action == Action::Jump && !self.jumping {
            self.jumping = true;
            self.dino_vy = INITIAL_JUMP_VELOCITY;
        }

        if self.jumping {
            self.dino_y += self.dino_vy;
            self.dino_vy += GRAVITY;
            if self.dino_y >= GROUND_Y {
                self.dino_y = GROUND_Y;
                self.dino_vy = 0.0;
                self.jumping = false;
            }
        }

        for o in self.obstacles.iter_mut() {
            o.x -= self.speed;
        }
        self.obstacles.retain(|o| o.x + o.w > -50.0);

        let rightmost = self.obstacles.iter().map(|o| o.x).fold(f32::MIN, f32::max);
        if self.obstacles.is_empty() || CANVAS_W - rightmost >= self.next_spawn_gap {
            self.spawn_obstacle();
        }

        if self.collides() {
            self.crashed = true;
            return false;
        }

        self.distance_ran += self.speed;
        if self.speed < MAX_SPEED {
            self.speed += ACCELERATION;
        }
        self.frame += 1;
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn replays_are_deterministic() {
        let run = |seed| {
            let mut g = Game::new(seed);
            let mut acts = vec![];
            let mut r = Rng::new(7);
            while g.tick(if r.next_f32() < 0.02 { Action::Jump } else { Action::Run }) {
                acts.push(g.distance_ran);
                if g.frame > 5_000 {
                    break;
                }
            }
            (g.frame, g.score(), acts.len())
        };
        assert_eq!(run(42), run(42));
    }

    #[test]
    fn different_seeds_diverge() {
        // Note: time-to-first-crash is *not* a valid signal here — the opening
        // obstacle always spawns at the right edge, so every seed dies on the
        // same frame if you never jump. Compare the generated layout instead.
        let layout = |seed| {
            let mut g = Game::new(seed);
            let mut kinds = vec![];
            for _ in 0..3_000 {
                g.crashed = false;
                g.tick(Action::Run);
                if let Some(o) = g.obstacles.last() {
                    if kinds.last() != Some(&(o.kind, o.w as i32)) {
                        kinds.push((o.kind, o.w as i32));
                    }
                }
            }
            kinds
        };
        assert_ne!(layout(1), layout(2), "seeds should produce different obstacle layouts");
    }

    #[test]
    fn standing_still_eventually_crashes() {
        let mut g = Game::new(3);
        let mut frames = 0;
        while g.tick(Action::Run) {
            frames += 1;
            assert!(frames < 10_000, "never hit an obstacle while standing still");
        }
        assert!(g.crashed);
    }

    #[test]
    fn a_jump_returns_to_ground() {
        let mut g = Game::new(9);
        g.tick(Action::Jump);
        assert!(g.jumping);
        let mut frames = 0;
        while g.jumping && g.tick(Action::Run) {
            frames += 1;
            assert!(frames < 200, "jump never landed");
        }
        // ~33 frames of hang time at gravity 0.6 / v0 -10.
        assert!((30..40).contains(&frames), "unexpected hang time: {frames}");
    }

    #[test]
    fn speed_ramps_to_cap_and_holds() {
        let mut g = Game::new(5);
        for _ in 0..20_000 {
            g.crashed = false; // ignore collisions; we only care about the ramp
            g.tick(Action::Run);
        }
        assert!((g.speed - MAX_SPEED).abs() < 1e-3, "speed = {}", g.speed);
    }
}
